ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs

```diff
@@ -551,14 +560,23 @@ peerSelectionGovernorLoop tracer
       now <- getMonotonicTime
       let Decision { decisionTrace, decisionJobs, decisionState } =
             timedDecision now
-          !newCounters = peerStateToCounters decisionState
+
+      mbCounters <- atomically $ do
+        -- Update counters
+        counters <- readTVar countersVar
+        let !counters' = peerSelectionStateToCounters decisionState
+        if counters' /= counters
+          then writeTVar countersVar counters'
+            >> return (Just counters')
+          else return Nothing
+
+      -- Trace counters
+      traverse_ (traceWith countersTracer) mbCounters
+
       traverse_ (traceWith tracer) decisionTrace
-      traceWithCache countersTracer
-                     (countersCache decisionState)
-                     newCounters
 
```

Comment:
Maybe we could rethink how we trace these things. Overall, if my understanding is correct, each decision changes only a few things in the state, but here it looks like we re-compute everything to trace. Maybe we could only recompute only the things that change, and the tracer could take the parts that did not change and add those parts that we had to re-calculate.

Reply:
But we don't know which things have changed, so we'll need to recompute them anyway.  But maybe there's something clever I am missing.

Reply:
Just thinking out loud, but maybe the job/monitoring action that records a change could return some sort of lens that references which fields it has touched, along with the entire counters record (most of which would remain unevaluated). Then code here could apply this lens to left and right hand sides just to see if there is indeed a difference, and also use the lens to update just the changed counters in countersVar.

Reply:
A fancy idea: we could use `NoThunks` (or its approach) to build a monoid that combines two records, picking the evaluated thunks.

Reply:
We have quite a few relationships between `PeerSelectionCounters` record fields which we'd need to preserve to really avoid recomputing things.  This would require some fancy machine to derive the right instance - might be a nice challenge.  Here's a [basic idea](https://gist.github.com/coot/b4bf403f858f5d3d6944eec90080b961) which seems to work when there are no dependencies between fields.

Testing also will be non trivial, because constructed record will not be without thunks (no `NoThunks` cannot be used), simply because this will require `Generics` and `to` will leave us with fields which are thunks themselves (although they will point to computed values).

Reply:
btw, I don't think that the `nothunks` approach is something we'd like to include in production because it relies on `unsafePerformIO` and some rather fancy `GHC` runtime API.  If we have a performance regression which cannot be accepted, we'll look into how to cache part of the data structure as you propose.  I hope that this won't be a problem because these structures are rather small (e.g. less than 1000 entries and in standard configuration even less than 100).

Reply:
The method you propose is very clever, but as you say maybe it won't be a problem. We could create a separate issue to perform some benchmarks to convince ourselves that the impact is indeed negligible, otherwise we can try to resolve it separately.

Reply:
https://github.com/IntersectMBO/ouroboros-network/issues/4857